language/smt.py

```diff
@@ -0,0 +1,139 @@
+from bl.BLVisitor import BLVisitor
+from bl.BLListener import BLListener
+from bl.BLParser import BLParser
+from antlr4 import *
+
+
+class DeclarationVisitor(BLVisitor):
+    def defaultResult(self) -> set[str]:
+        return set()
+
+    def aggregateResult(self, aggregate: set[str], nextResult: set[str]):
+        if len(aggregate.intersection(nextResult)) > 0:
+            raise ValueError(
+                f"repeated definition of variables {aggregate.intersection(nextResult)}"
+            )
+
+        return aggregate.union(nextResult)
+
+    def visitAssign(self, ctx: BLParser.AssignContext):
+        variable_name = ctx.children[1].symbol.text
+        return {variable_name}
+
+
+class SMTVisitor(BLVisitor):
+    def defaultResult(self):
+        return ""
+
+    def aggregateResult(self, aggregate: str, nextResult: str):
+        if len(aggregate) == 0 or len(nextResult) == 0:
+            return aggregate + nextResult
+        return aggregate + " " + nextResult
+
+    def visitAssigns(self, ctx: BLParser.AssignsContext) -> str:
+        return self.visitChildren(ctx)
+
+    def visitExpr(self, ctx: BLParser.ExprContext):
+        if len(ctx.children) == 1:
+            return self.visitChildren(ctx)
+
+        # expr -> expr '+' multexpr
+        # expr -> expr '-' multexpr
+
+        if ctx.children[1].getText() == "+":
+            text = "(+ "
+        elif ctx.children[1].getText() == "-":
+            text = "(- "
+        else:
+            assert False
+
+        text += self.visitChildren(ctx)
+        text += ")"
+
+        return text
+
+    def visitMultexpr(self, ctx: BLParser.MultexprContext):
+        if len(ctx.children) == 1:
+            return self.visitChildren(ctx)
+
+        # expr -> expr '*' multexpr
+        # expr -> expr '/' multexpr
+
+        if ctx.children[1].getText() == "*":
+            text = "(* "
+        elif ctx.children[1].getText() == "/":
+            text = "(/ "
+        else:
+            assert False
```

Comment:
Is this line, and line 48, necessary? Wouldn't the case where an expression is neither `+` or `-` captured by the earlier pass?